consensus/istanbul/core/commit.go

```diff
@@ -207,62 +218,135 @@ func (c *core) handleCheckedCommitForCurrentSequence(msg *istanbul.Message, comm
 		return errInvalidValidatorAddress
 	}
 
-	if err := c.verifyCommittedSeal(commit, validator); err != nil {
-		return errInvalidCommittedSeal
+	// ensure that the commit is for the current proposal
+	if err := c.verifyCommit(commit); err != nil {
+		return err
 	}
 
 	newValSet, err := c.backend.NextBlockValidators(c.current.Proposal())
 	if err != nil {
 		return err
 	}
 
+	// Verifies the individual seal for this commit message for the epoch
+	// block, no-op unless this is the last block of an epoch.
 	if err := c.verifyEpochValidatorSetSeal(commit, c.current.Proposal().Number().Uint64(), newValSet, validator); err != nil {
 		return errInvalidEpochValidatorSetSeal
 	}
 
-	// ensure that the commit is in the current proposal
-	if err := c.verifyCommit(commit); err != nil {
-		return err
-	}
-
 	// Add the COMMIT message to current round state
 	if err := c.current.AddCommit(msg); err != nil {
 		logger.Error("Failed to record commit message", "m", msg, "err", err)
 		return err
 	}
-	numberOfCommits := c.current.Commits().Size()
+
+	commits := c.current.Commits()
+	numberOfCommits := commits.Size()
 	minQuorumSize := c.current.ValidatorSet().MinQuorumSize()
 	logger.Trace("Accepted commit for current sequence", "Number of commits", numberOfCommits)
 
 	// Commit the proposal once we have enough COMMIT messages and we are not in the Committed state.
 	//
-	// If we already have a proposal, we may have chance to speed up the consensus process
-	// by committing the proposal without PREPARE messages.
 	// TODO(joshua): Remove state comparisons (or change the cmp function)
 	if numberOfCommits >= minQuorumSize && c.current.State().Cmp(StateCommitted) < 0 {
-		logger.Trace("Got a quorum of commits", "tag", "stateTransition", "commits", c.current.Commits)
-		err := c.commit()
+		proposal := c.current.Proposal()
+		// TODO understand how proposal can be nil.
+		if proposal == nil {
+			return nil
+		}
+
+		// Generate aggregate seal
+		aggregatedSeal, err := c.generateAggregateCommittedSeal()
 		if err != nil {
-			logger.Error("Failed to commit()", "err", err)
-			return err
+			logger.Warn("Initial verificaction of aggregate commit signature failed", "err", err)
+			// Remove any bad committed seals and try again if sufficient commits remain
+			c.removeInvalidCommittedSeals()
+			if c.current.Commits().Size() < c.current.ValidatorSet().MinQuorumSize() {
+				// Wait for more commits
+				return nil
+			}
+			aggregatedSeal, err = c.generateAggregateCommittedSeal()
+		}
+		// If there is still an error then sit this round out, we can't continue.
+		if err != nil {
+			nextRound := new(big.Int).Add(c.current.Round(), common.Big1)
+			logger.Warn("Error on commit, waiting for desired round", "reason", "failed to aggregate commit seals", "err", err, "desired_round", nextRound)
```

Comment:
IIUC, this can only happen if `removeInvalidCommittedSeals` fails and puts us in a state where there are a quorum of commits, but they do not verify. If this is the case, we will want to log a louder error. Arguably, we might `panic`, but I'm not sure that's wise.

Reply:
So I was just maintaining the previous logic here, this is the case where there is something else wrong with the aggregation that is not an invalid signature. I would consider this a programming error and I would be inclined to panic here but I didn't want to go tweaking this unrelated logic as part of this PR.

Reply:
I think your changes make this more obviously a programming error if it happens, and I think we should raise this message to an error level right now.
```suggestion
			logger.Error("Error on commit, waiting for desired round", "reason", "failed to aggregate commit seals", "err", err, "desired_round", nextRound)
```

Reply:
I've updated this and a few similar cases to use Error